synapse/storage/registration.py

```diff
@@ -668,11 +668,21 @@ def get_threepid_validation_session(
                 perform no filtering
 
         Returns:
-            deferred {str, int}|None: A dict containing the
-                latest session_id and send_attempt count for this 3PID.
-                Otherwise None if there hasn't been a previous attempt
+            deferred {str, str|int|None}|None: A dict containing the following:
+                * address - address of the 3pid
+                * medium - medium of the 3pid
+                * client_secret - a secret provided by the client for this validation session
+                * session_id - ID of the validation session
+                * send_attempt - a number serving to dedupe send attempts for this session
+                * validated_at - timestamp of when this session was validated if so
+
+                Otherwise None if a validation session is not found
         """
-        keyvalues = {"medium": medium, "client_secret": client_secret}
+        keyvalues = {}
+        if client_secret:
```

Comment:
I'd suggest `if client_secret is not None` here to avoid any thinkos where you can set the client_secret to "" to validate someone else's email address.

(Indeed, is doing a validation without a client_secret and/or a sid something that we should allow?)

Reply:
We always want a `client_secret`, however we don't always have a session id: https://github.com/matrix-org/synapse/blob/9fc71dc5eed7531454a34f8fec34bd451458c7c6/synapse/handlers/identity.py#L315

I'll make client_secret required though.